pkg/util/pki/nameconstraints.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2023 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pki
+
+import (
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"net"
+
+	v1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+)
+
+// Copied from x509.go
+var (
+	OIDExtensionNameConstraints = []int{2, 5, 29, 30}
+)
+
+// NameConstraints represents the NameConstraints extension.
+type NameConstraints struct {
+	PermittedDNSDomainsCritical bool        `asn1:"optional,explicit,tag:0"`
```

Comment:
I missed this in the review, but this is not correct, this is not what is described in the x509 spec.
/cc @maelvls

Reply:
@tanujd11 could you create a new PR to fix this?

Reply:
Hey @inteon , what exactly are you referring here?

Reply:
The marshalled NameConstraints extension in the CSR has to match the format defined in the RFC (see https://www.alvestrand.no/objectid/2.5.29.30.html).
The encoding you specified seems to be custom.
We can use the logic the x509 library uses: https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/crypto/x509/x509.go;l=1189-1285;drc=82c713feb05da594567631972082af2fcba0ee4f